server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -646,7 +628,43 @@ private static SnapshotsInRepo sortSnapshots(
         final List<SnapshotInfo> resultSet = size != GetSnapshotsRequest.NO_LIMIT && size < snapshots.size()
             ? snapshots.subList(0, size)
             : snapshots;
-        return new SnapshotsInRepo(resultSet, filteredSnapshotInfos.size(), allSnapshots.size() - resultSet.size());
+        return new SnapshotsInRepo(resultSet, snapshotInfos.size(), allSnapshots.size() - resultSet.size());
+    }
+
+    private static List<SnapshotInfo> filterBySLMPolicies(List<SnapshotInfo> snapshotInfos, String[] slmPolicies) {
+        final List<String> includePatterns = new ArrayList<>();
+        final List<String> excludePatterns = new ArrayList<>();
+        boolean seenWildcard = false;
+        for (String slmPolicy : slmPolicies) {
+            if (seenWildcard && slmPolicy.length() > 1 && slmPolicy.startsWith("-")) {
+                excludePatterns.add(slmPolicy.substring(1));
+            } else {
+                if (Regex.isSimpleMatchPattern(slmPolicy)) {
+                    seenWildcard = true;
+                }
+                includePatterns.add(slmPolicy);
+            }
+        }
+        final String[] includes = includePatterns.toArray(Strings.EMPTY_ARRAY);
+        final String[] excludes = excludePatterns.toArray(Strings.EMPTY_ARRAY);
+        return snapshotInfos.stream().filter(snapshotInfo -> {
+            final Map<String, Object> metadata = snapshotInfo.userMetadata();
+            final String policy;
+            if (metadata == null) {
+                policy = "";
```

Comment:
I think we have a choice here. Since there could be no slm-policy (in contrast to repo and name), we could choose between a few options:

1. `*` also matches no policy (like here). This allows `*,-a*` to give all snapshots with no policy starting with `a` (including all those with no policy), but no easy way to find all those with no policy.
2. `*` only matches those that have a policy. This allows `*` to find all those with a policy and we could define `-*` to find all those without a policy.

So 2 seems more flexible but deviates from how we handle repo and name. But then again, we are defining this one from scratch so we have the freedom. Let me know your thoughts.

Reply:
Hmm I think I like `2` better as well. I wonder though, the `-*` trick we'll have to document anyway. Maybe we should use a special parameter `_none` (like we have for the ingest pipeline name to turn off a pipeline) for this. That way we could even do queries that involve a set of policies + all snapshots without any policies (though we could do this with this pattern as well, it seem incredibly confusing to have this as the only exclude pattern that is technically a positive match pattern for `null`)?

Reply:
++ to using `_none` as a special `null` value.

Reply:
+1 to the null value pattern, too